dbcon/mysql/ha_mcs_opt_rewrites.cpp

```diff
@@ -243,3 +246,86 @@ simplify_joins_mcs(JOIN *join, List<TABLE_LIST> *join_list, COND *conds, bool to
   }
   DBUG_RETURN(conds);
 }
+
+/*@brief  in_subselect_rewrite_walk - Rewrites Item_in_subselect*/
+/************************************************************
+* DESCRIPTION:
+* It traverses filter predicates searching for
+* Item_in_subselect and rewrites it adding equi-join predicate
+* to finalise IN_2_EXISTS rewrite.
+* PARAMETERS:
+*    item_arg - Item to check.
+*    arg - unused dummy arg.
+* RETURN:
+***********************************************************/
+void in_subselect_rewrite_walk(const Item* item_arg, void* arg)
+{
+    bool* result= reinterpret_cast<bool*>(arg);
+    if (*result) return;
+
+    Item* item= const_cast<Item*>(item_arg);
+
+    JOIN* join= nullptr;
+    if (typeid(*item) == typeid(Item_in_subselect))
+    {
+        Item_in_subselect* sub= reinterpret_cast<Item_in_subselect*>(item);
+        // MCS 1.4.3 doesn't support IN + subquery with UNION so
+        // we safe to take this JOIN.
+        join= sub->unit->first_select()->join;
```

Comment:
Curious, how does this logic work for the case of an IN + derived subquery? e.g.:
select field1 from t1 where field2 in (select * from (select field3 from t2) a);

Reply:
There will be no join inside Item_in_subselect that describes IN subquery so we just inject equi-join predicate into IN subquery.

Reply:
This is good to know.